src/coordinate-grid-mixin.js

```diff
@@ -1231,85 +1230,108 @@ dc.coordinateGridMixin = function (_chart) {
+                });
+            }
+
+            _chart._invokeZoomedListener();
             dc.events.trigger(function () {
-                _rangeChart.replaceFilter(domFilter);
-                _rangeChart.redraw();
-            });
+                _chart.redrawGroup();
+            }, dc.constants.EVENT_DELAY);
         }
-
-        _chart._invokeZoomedListener();
-
-        dc.events.trigger(function () {
-            _chart.redrawGroup();
-        }, dc.constants.EVENT_DELAY);
-
-        _refocused = !rangesEqual(domain, _xOriginalDomain);
     }
 
-    function onZoom () {
-        if (!d3.event.sourceEvent && d3.event.sourceEvent.type !== 'zoom') { return; }
+    // Our zooming is not standard d3 zoom as defined in their examples.
+    // Instead we want to focus the chart based on current zoom transform.
+    // The following code computes values of new domain the same way as transform.rescaleX.
+    // The difference is what we do after we get the newDomain
+    var _zoomTransformToDomain = function (transform, xScale) {
+        return xScale.range().map(function (xCoord) {
+            return _origX.invert(transform.invertX(xCoord));
+        });
+    };
 
-        _chart.x(d3.event.transform.rescaleX(_origX));
+    // _zoomTransformToDomain(transform, xScale) should give back newDomain
+    var _domainToZoomTransform = function (newDomain, origDomain, xScale) {
+        var k = (origDomain[1] - origDomain[0]) / (newDomain[1] - newDomain[0]);
+        var xt = -1 * xScale(newDomain[0]);
 
-        zoomHandler();
-    }
+        return d3.zoomIdentity.scale(k).translate(xt, 0);
+    };
 
-    function intersectExtents (ext1, ext2) {
-        if (ext1[0] > ext2[1] || ext1[1] < ext2[0]) {
-            console.warn('could not intersect extents');
+    // If we changing zoom status (for example by calling focus), tell D3 zoom about it
+    var _updateD3zoomTransform = function () {
+        if (_zoom) {
+            _zoom.transform(_chart.root(), _domainToZoomTransform(_chart.x().domain(), _xOriginalDomain, _origX));
         }
-        return [Math.max(ext1[0], ext2[0]), Math.min(ext1[1], ext2[1])];
-    }
+    };
 
-    function constrainExtent (extent, constraint) {
-        var size = extent[1] - extent[0];
-        if (extent[0] < constraint[0]) {
-            return [constraint[0], Math.min(constraint[1], dc.utils.add(constraint[0], size, 'millis'))];
-        } else if (extent[1] > constraint[1]) {
-            return [Math.max(constraint[0], dc.utils.subtract(constraint[1], size, 'millis')), constraint[1]];
-        } else {
-            return null;
-        }
+    function onZoom () {
+        var event = d3.event;
+        // Avoids infinite recursion
+        // To ensure that when it is called because of programatic zoom there is no d3.event.sourceEvent
+        d3.event = null;
+        if (!event.sourceEvent) { return; }
+
+        var newDomain = _zoomTransformToDomain(event.transform, _origX);
```

Comment:
This seems to work afaict:

```js
        var newDomain = event.transform.rescaleX(_origX).domain();
```

Reply:
I think it would. :) The more I am getting to know D3 the more I admire.

I reached this code independently step at a time (using drawings on a paper with pencil). Just now rechecked - this code seems to do exactly the same. Let me know if you would change or should I.

Reply:
Okay, great. I am working on it. 

I hope to also clean up the recursive event stuff and merge this today. (The `zoomHandler` argument makes sense to me, but setting `d3.event` seems messy so I'm looking at using the same strategy there.)

Reply:
The `d3.event`, I could not understand their reasoning behind keeping the event in a global variable. I think it would have been much cleaner design for D3 authors as well as users if D3 passed the event as parameter to event handler (or, it already does and I am unaware).

I had started setting it to `null` as I was getting occasional infinite loop. Maybe those were caused by something else. D3 examples don't do that.

Reply:
I don't understand the global either, and I don't remember seeing a rationale for it. It is what it is. :wink:

Yes, it's easy to repro the infinite recursion. I'm experimenting with using arguments similar to `noRaiseEvents` to fix this. Still makes me think that we might have things inside-out, at least for the new way that `d3.zoom` works. 

However, it's likely that the problem is abstraction itself: often things that are easy in ad-hoc UI code become really complicated when you try to make the code reusable. A lot of people reject dc.js for this reason, but I still think it's worth it!

Reply:
Actually I am not sure about others, if it was not for dc, I would not dare to build the functionality that I need using D3 directly. D3 is like an ocean for a new person.

So definitely worth it.